crates/glue/src/types.rs

```diff
@@ -111,24 +111,21 @@ impl Types {
             target,
         );
 
-        let variables: Vec<_> = entry_points.values().copied().collect();
-        for var in variables {
+        for (symbol, var) in entry_points.into_iter() {
             env.lambda_set_ids = env.find_lambda_sets(var);
-            let id = env.add_toplevel_type(var, &mut types);
 
-            let key = entry_points
-                .iter()
-                .find_map(|(k, v)| (*v == var).then_some((*k, id)));
+            let id = env.add_toplevel_type(var, &mut types);
+            let name = symbol.as_str(env.interns).to_string();
 
-            if let Some((k, id)) = key {
-                let name = k.as_str(env.interns).to_string();
+            if is_function(env.subs, var) {
                 types.entry_points.push((name, id));
-                entry_points.remove(&k);
+            } else {
+                // This top-level value isn't a function, so we will end up code-genning a thunk
+                // for it. We need to represent that thunk explicitly in glue!
+                env.add_thunk_entry_point(&mut types, name, id);
             }
```

Comment:
I'm not sure this logic is correct. Wouldn't the following program run into the same issue?

```
mainImpl1 = \{} -> ""
mainImpl2 = \{} -> ""

mainForHost : {} -> Str
mainForHost = if Bool.true then mainImpl1 else mainImpl2
```

This would hit the first branch on 120, but `mainForHost` should also be a thunk here.